sequencefile writer

kitchen opened the pull request:
adds a bunch of new functions:
- `WriteVInt` writes a vint value
- `WriteString` which writes a string out prefixed with a vint saying what the length is
- `readBoolean` and `writeBoolean` which read and write boolean values (per java's serialization of these)
- `PutsWritable` takes a `[]byte` and turns it into a BytesWritable-wrapped `[]byte` (which is just prefixed with 4 bytes representing the uint32 length of the byte array)

It then builds on those functions and writes out the sequencefile header, followed by a sync marker,   and provides functions for putting in key/value pairs.

still remaining to be done: block compression and placing sync markers every so often. Also, there's very little in the way of sanity checking being done here, we may want to add some validation to things.

kitchen committed:
add a WriteVInt function

kitchen committed:
add tests for WriteVInt

they currently fail, but that's fine, that means something is wrong!

kitchen committed:
copy/paste error with i, idx, n from java

kitchen committed:
remove extra ()

kitchen committed:
separate tests for ReadVInt and WriteVInt

kitchen committed:
don't mask top-level `len`

kitchen committed:
only one call to Write

kitchen committed:
return the number of bytes written

kitchen committed:
remove extra int()

kitchen committed:
return directly

kitchen committed:
clarify

kitchen committed:
shift and assign, don't just shift

don't know how this was passing tests before, but it was.

kitchen committed:
gofmt fixes

kitchen committed:
add some docs for WriteVInt

kitchen committed:
add a basic Writer type and WriteString function (and tests)

kitchen committed:
wrap each string into a test run

kitchen committed:
readBoolean and writeBoolean

kitchen committed:
writemetadata and tests

kitchen committed:
change syncmarker on Header to a byte slice instead of string

kitchen committed:
writeSyncMarker()

kitchen committed:
sort the metadata before writing

this both fixes the test because of golang's correct behavior of random
ordering when doing a range on a map, and makes the output
deterministic.

kitchen committed:
write the full header (and tests)

kitchen committed:
better defaults, compression stuff

kitchen committed:
add a couple of TODOs

kitchen committed:
start writing keys/values

kitchen committed:
factor out PutsWritable and leave serialization up to the user

kitchen committed:
test multiple pairs, make sure buffer is empty when we're done

kitchen committed:
this should return a decompressor

kitchen committed:
break a value up into multiple snappy frames of 128KB size

kitchen committed:
also test a very long string of "a" repeating

I added some output into the chunker and wasn't getting very good
compression but assumed that was mostly due to random data not
compressing well, so wanted to add a case that *should* compress well
(it did)

kitchen committed:
refactor tests a bit

colinmarc commented on header.go:
```diff
@@ -73,11 +77,12 @@ func (r *Reader) ReadHeader() error {
 			return err
 		}
 
+		// TODO: DRY this out along with other COMPRESSION
```
Leaving a comment so you don't forget this (I think it's probably fine to just leave the duplication)

colinmarc commented on header.go:
```diff
@@ -98,13 +103,82 @@ func (r *Reader) ReadHeader() error {
 	}
 
-	r.Header.SyncMarker = string(marker)
+	r.Header.SyncMarker = marker
```
You need to copy this, or the byte slice will be reused (that's how `consume` works). I'm worried that this didn't break tests because we only have tiny sequence files.

kitchen replied:
we don't just have tiny sequencefiles in tests, my round trip tests write out rather large sequence files. And I just added some more, smaller keys, but more of them. I did address the issue though!

colinmarc commented on header.go:
```diff
@@ -98,13 +103,82 @@ func (r *Reader) ReadHeader() error {
 	}
 
-	r.Header.SyncMarker = string(marker)
+	r.Header.SyncMarker = marker
```
Oh, I see we still have `syncMarkerBytes`, which is actually a copy. This'll still be incorrect as soon as we read anything else, though.

colinmarc commented on header.go:
```diff
@@ -136,6 +210,63 @@ func (r *Reader) readMetadata() error {
+	totalwritten := 0
+
+	length := len(w.Header.Metadata)
+	lengthbytes := make([]byte, 4)
```
`lengthBytes`

colinmarc commented on header.go:
```diff
@@ -136,6 +210,63 @@ func (r *Reader) readMetadata() error {
+	lengthbytes := make([]byte, 4)
+	binary.BigEndian.PutUint32(lengthbytes, uint32(length))
+	written, err := w.writer.Write(lengthbytes)
+	totalwritten += written
```
`totalWritten` (and elsewhere)

colinmarc commented on vint.go:
```diff
@@ -60,3 +60,37 @@ func mustReadByte(r io.Reader) (byte, error) {
+
+// WriteVInt writes an int64 encoded in hadoop's "VInt" format, described and
+// implemented here: https://goo.gl/1h4mrG. It does at most one write to the
+// underlying io.Writer
```
period.

colinmarc commented on writer.go:
```diff
@@ -0,0 +1,166 @@
+	sinceLastSync int
+	compressor    compressor
+}
+
```
Let's add a `Create` method that takes a path, a la, `Open`, and wraps the writer in a `bufio.Writer` (and remembers to flush at the end).

colinmarc commented on writer.go:
```diff
@@ -0,0 +1,166 @@
+	compressor    compressor
+}
+
+func NewWriter(w io.Writer) *Writer {
```
This needs a godoc, and the godoc should encourage you to wrap `w` in a `bufio.Writer`, so that the individual writes don't suck.

colinmarc commented on writer.go:
```diff
@@ -0,0 +1,166 @@
+func NewWriter(w io.Writer) *Writer {
+	header := Header{
+		Version:        6,
+		KeyClassName:   "org.apache.hadoop.io.BytesWritable",
```
https://github.com/colinmarc/sequencefile/blob/master/writable.go#L10

kitchen replied:
I agree. I'm currently operating under the "just modify the underlying `Header` like I am doing in the tests with compression bits.

colinmarc commented on writer.go:
```diff
@@ -0,0 +1,166 @@
+func NewWriter(w io.Writer) *Writer {
+	header := Header{
+		Version:        6,
+		KeyClassName:   "org.apache.hadoop.io.BytesWritable",
```
Also this should be a setting of some sort.

colinmarc commented on writer.go:
```diff
@@ -0,0 +1,166 @@
+	return &Writer{writer: w, Header: header, sinceLastSync: 0}
+}
+
+func (w *Writer) Write(bytes []byte) (int, error) {
```
godocs!

kitchen replied:
I think it's just a remnant of another thought process and can go away

colinmarc commented on writer.go:
```diff
@@ -0,0 +1,166 @@
+	return &Writer{writer: w, Header: header, sinceLastSync: 0}
+}
+
+func (w *Writer) Write(bytes []byte) (int, error) {
```
What's the point of this method, btw?

kitchen committed:
track amount written and actually write sync marker

also the sync marker is FFFFFFFF followed by SyncSize bytes of random,
not just the SyncSize bytes of random, so write that out, too.

colinmarc commented on writer.go:
```diff
@@ -0,0 +1,166 @@
+	return w.writer.Write(w.Header.SyncMarker)
+}
+
+// maybe
```
that's not how godocs are supposed to look

colinmarc commented on writer.go:
```diff
@@ -0,0 +1,166 @@
+}
+
+func (w *Writer) Append(key []byte, value []byte) (int, error) {
+	// TODO: if we haven't written the header yet, should we error, or just silently write the header?
```
don't do either - we should let people write streams without a header.

colinmarc commented on writer.go:
```diff
@@ -0,0 +1,174 @@
+
+func NewWriter(w io.Writer) *Writer {
+	header := Header{
+		Version:        6,
```
this should be a constant

colinmarc commented on writer.go:
```diff
@@ -0,0 +1,174 @@
+		ValueClassName: "org.apache.hadoop.io.BytesWritable",
+		Compression:    NoCompression,
+	}
+	return &Writer{writer: w, Header: header, sinceLastSync: 0}
```
You don't need `sinceLastSync: 0`, `0` is the zero value for int.

kitchen replied:
:yayoctopus:

colinmarc commented on writer.go:
```diff
@@ -0,0 +1,174 @@
+	}
+
+	keylength := len(key)
+	keylengthbytes := make([]byte, 4)
```
`camelCase` these

colinmarc commented on writer.go:
```diff
@@ -0,0 +1,174 @@
+	recordlengthbytes := make([]byte, 4)
+	binary.BigEndian.PutUint32(recordlengthbytes, uint32(recordlength))
+
+	written, err = w.writer.Write(recordlengthbytes)
```
I think we should refactor this such that we're only doing one write to the underlying writer for each call to `Append` or `WriteHeader`, even if that's slightly less optimal for the case where the underlying writer _is_ a buffer of some sort.

Another alternative would be to write two methods:

``` go
func (w *Writer) write(b []byte) (int, error) {
    if bw, ok := w.writer.(*bufio.Writer); ok {
        return bw.Write(b)
    }

    return w.buf.Write(b) // this is a bytes.Buffer
}

func (w *Writer) flush(b []byte) error {
    if bw, ok := w.writer.(*bufio.Writer); ok {
        return nil
    }

    _, err := io.Copy(w.writer, w.buf)
    return err
}
```

You could also think about tracking `totalWritten` generically in there.

colinmarc replied:
Per PM, if we know we're getting a `Flush` method with block writing anyway, then let's just hook it up to that and call it a day.

kitchen replied:
except the underlying Writer is an `io.Writer` which doesn't specify a `Flush()`

is there a way to selectively check for that in our `Flush()`?

kitchen replied:
or are you saying just default to wrapping the underlying Writer with a `bufio.Writer` and call flush on _that_? That could work.

colinmarc replied:
Yeah, exactly.

kitchen replied:
I still need to address this (notes for me for later)

colinmarc commented on writer.go:
```diff
@@ -0,0 +1,174 @@
+	recordlengthbytes := make([]byte, 4)
+	binary.BigEndian.PutUint32(recordlengthbytes, uint32(recordlength))
+
+	written, err = w.writer.Write(recordlengthbytes)
```
(Note that that alternative works for `bufio.Writer`s, which we'd use if you just used `Create`, but not for `bytes.Buffer`, `io.Pipe`, etc)

colinmarc commented on writer.go:
```diff
@@ -0,0 +1,174 @@
+	}
+}
+
+func (w *Writer) compressSnappy(raw []byte) ([]byte, error) {
```
It'd be better to have an interface type for this, like we do with `decompressor`

kitchen replied:
yea, I was trying to. I think it's harder because of the framing format. and because blockcompression. It's been a while since I thought about this though so I'll look it over again and think about it and see :)

colinmarc commented on writer.go:
```diff
@@ -0,0 +1,174 @@
+	binary.BigEndian.PutUint32(rawlenbytes, uint32(rawlen))
+
+	ret := make([]byte, 0, 4)
+	ret = append(ret, rawlenbytes...)
```
You can just `PutUint32` directly into `ret`

colinmarc commented on writer.go:
```diff
@@ -0,0 +1,174 @@
+		}
+
+		tmp := make([]byte, snappy.MaxEncodedLen(len(chunk))+4)
+		encoded_chunk := snappy.Encode(tmp, chunk)
```
`encodedChunk`

colinmarc commented on writer.go:
```diff
@@ -0,0 +1,174 @@
+
+		encoded_chunklen := len(encoded_chunk)
+		encoded_chunklenbytes := make([]byte, 4)
+		binary.BigEndian.PutUint32(encoded_chunklenbytes, uint32(encoded_chunklen))
```
again here - you've a few too many slice allocations

kitchen replied:
do I, though? I have to capture the encoded chunk by itself so I can determine its length, then I need to `PutUint32` that into something, then append both of those to ret. And `snappy.Encode` needs a workspace, so `tmp` it is.

maybe? :D

kitchen committed:
WriteHeader and Append don't return bytes written anymore

beacuse Append doesn't actually write to the underlying writer directly
every time if we're doing blockcompression

kitchen committed:
PutVInt function

colinmarc commented on vint.go:
```diff
@@ -65,8 +65,14 @@ func mustReadByte(r io.Reader) (byte, error) {
+
+// PutVInt returns a byte slice containing an int64 encoded in hadoop's "VInt" format,
+// described and implemented here: https://goo.gl/1h4mrG
+func PutVInt(n int64) []byte {
```
I'd make this unexported - it's a bit of an odd method.

kitchen replied:
is there a way to make it less odd that you could see being exported?

the vint stuff I can kinda see as eventually getting extracted out as a generic library, along with *writable and such. Like a "hadoop data format library" ... which would also mirror the java implementation that does the same thing. I think in that case PutVint would have to be exported for sequencefile to be able to use it, yes?

colinmarc commented on writable.go:
```diff
@@ -18,6 +18,13 @@ func BytesWritable(b []byte) []byte {
 	return b[4:]
 }
 
+func PutBytesWritable(raw []byte) []byte {
```
We're missing `PutText`, `PutIntWritable`,  etc. Also, for consistency with `encoding/binary`'s `Put*` methods, I think it'd be better to pass in the byte slice.

kitchen replied:
so in the case of `encoding/binary`'s `Put*` methods, you pass in the byte slice but it already has to be of a size large enough to hold the value being placed in: https://github.com/kitchen/gokitchentest/blob/master/putuint32_test.go but those are also all fixed length representations, so you can know them in advance. However, I don't know if that is the case with the other Writables, or if that's even an expected property of them. (I also don't understand exactly why `encoding/binary` blows up if you give it a byte slice that isn't already the correct size, even if it has been allocated to be large enough.)

kitchen replied:
oh, so I guess the question I missed there was: should it just accept a byte slice and yolo the first X bytes, where X is whatever number of bytes it needs?

kitchen committed:
blockcompression works

kitchen committed:
copy syncmarker slice

kitchen committed:
style changes

kitchen committed:
make the version number a constant

kitchen committed:
zero value for int is 0

imagine that

kitchen committed:
PutUint32 directly into ret

wwkeyboard commented:
Is this branch still in progress? This would be very helpful for a project I'm working on.

kitchen commented:
it's feature complete but could use some cleanup, especially around the interface(s) and such.

I am unlikely to continue working on it any time soon, however. @colinmarc might have ideas for how to get it ready for merging if you wanted to pick it up!

kai5263499 commented:
Did this get anywhere? Writing would be very helpful for a project I'm working on as well

colinmarc commented:
Fixed in #5.
